extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/DataSourcePlan.java

```diff
@@ -413,15 +418,25 @@ private static DataSourcePlan forLookup(
   private static DataSourcePlan forQuery(
       final QueryKitSpec queryKitSpec,
       final QueryDataSource dataSource,
+      final QueryContext queryContext,
```

Comment:
Why is query context required ?

Reply:
The changes of this file are essentially to pass the `WINDOW_FUNCTION_OPERATOR_TRANSFORMATION` flag to the `WindowOperatorQueryKit` layer (as otherwise the flag set in `MSQTaskQueryMaker` layer isn't passed to `WindowOperatorQueryKit`).

So, for that, I needed to set the flag in the `query` being passed in `queryKitSpec.getQueryKit().makeQueryDefinition()` call in this method.

So I've passed the query context to this method to be able to get the flag's value.

Reply:
https://github.com/implydata/druid/blob/e5be9f13ba674e2f7780dc90ee72e6056429de50/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java#L577
Why don't we pass it here ? Like let the queryKit have a flag ?

Reply:
I'm not sure why we don't pass query context through the queryKit layers directly. We can certainly add a `queryContext` param in `QueryKit#makeQueryDefinition` interface method, and it'll help clean up the changes of DataSourcePlan layer.

I don't have a strong preference either way on this, happy to make the change if you'd prefer that.

(Although on second thoughts, a better way to do it might be to take it as a separate work item, and also check if some existing logic of DataSourcePlan layer also becomes unnecessary after passing query context directly - not sure if that might be the case though)

Reply:
Its confusing. 
As a developer, dataSource.getQuery() should have all the data that is required for me to take a decision. Like the most obvious question is how is datasource.getQuery().getCOntext() is different from queryContext. That nuance is hard to reason about when a person is writing an implementation.

Reply:
> Like the most obvious question is how is datasource.getQuery().getCOntext() different from queryContext. That nuance is hard to reason about when a person is writing an implementation.

@cryptoe Yeah, I fully agree, I was also confused on this.

The problem seems to be that wherever we do `query.withOverriddenContext(some context overrides)`, the context overrides are applied only to the top-level-query, not to all the nested `Query` objects within the query-datasource-query-chains.

I'm not sure if the above is the "only" thing leading to the mismatched context maps, but it's definitely one of the reasons.

I'm not sure if it's intentional, but maybe it is? For example, we add a `scanSignature` key to ScanQuery object, and it doesn't make sense to pass it to all the nested Query objects. So maybe the design originally stemmed from such requirements? (Not sure though, I'm lacking the historic context behind this design choice)

Thoughts?

Reply:
> https://github.com/implydata/druid/blob/e5be9f13ba674e2f7780dc90ee72e6056429de50/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java#L577
Why don't we pass it here ? Like let the queryKit have a flag ?

@cryptoe As discussed offline, have made this change.